### PR TITLE
[SPARK-19263] Fix race in SchedulerIntegrationSuite.

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?

All the process of offering resource and generating `TaskDescription` should be guarded by taskScheduler.synchronized in `reviveOffers`, otherwise there is race condition.

## How was this patch tested?

Existing unit tests.

### CODE REVIEW

**Comment by jinxing64** [comment]
@kayousterhout @squito 
This is originally raised by @squito when review https://github.com/apache/spark/pull/16620. Sorry for my eager to make this small pr.

**Comment by squito** [comment]
lgtm

sorry my fault in not breaking this out in the first place.

**Comment by jinxing64** [comment]
@squito
Thanks a lot for review.

File: core/src/test/scala/org/apache/spark/scheduler/SchedulerIntegrationSuite.scala [hunk 1]
```diff
@@ -391,17 +391,18 @@ private[spark] abstract class MockBackend(
...
-      }
-    }
-    synchronized {
+    // Need a lock on the entire scheduler to protect freeCores -- otherwise, multiple threads
+    // may make offers at the same time, though they are using the same set of freeCores.
+    taskScheduler.synchronized {
+      val newTaskDescriptions = taskScheduler.resourceOffers(generateOffers()).flatten
+      // get the task now, since that requires a lock on TaskSchedulerImpl, to prevent individual
+      // tests from introducing a race if they need it.
+      val newTasks =
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/SchedulerIntegrationSuite.scala]: super nit but can you fix the spacing here, so the newTaskDescriptions.map is on the previous line after the "=", and then the closing bracket on line 405 will line up with the next line.

**Comment by jinxing64** [comment]
@kayousterhout 
Thanks a lot for review. I've already refined.

**Review by kayousterhout — APPROVED** [review]
LGTM; I can merge this once tests pass

**Comment by jinxing64** [comment]
@kayousterhout  thanks a lot.
I'm not sure how to start the unit test automatically, do I have the right to do that?
BTW, may I ask a question, what is the proper way to run the unit test in my local?
Currently, I run a single unit test by `build/mvn test -DwildcardSuites=org.apache.spark.**Suite` and run all unit tests by ` build/mvn -Pyarn -Phadoop-2.6 -Phive-thriftserver -Dhadoop.version=2.6.5 test`. Is it correct?
It'd great if you can help me out or give a wiki that I can refer :)

**Comment by squito** [comment]
@jinxing64 that way of testing is fine, but I find its much faster to use sbt.

http://www.scala-sbt.org/0.13/docs/Testing.html

```
build/sbt -Pyarn -Phadoop-2.6 -Phive-thriftserver -Dhadoop.version=2.6.5
[this will put you in an sbt console]
> project core
> testOnly *DAGSchedulerSuite
[run all tests that match the pattern -- in this case, one suite]
> testOnly *spark.scheduler.*
[this time we run everything in the scheduler package]
>~testOnly *DAGSchedulerSuite
[the '~' in front means that as we modify the code (eg. in another terminal or an IDE), sbt will re-run the tests everytime the source changes.]
>~testOnly *DAGSchedulerSuite -- -z "SPARK-12345"
[as above, but only run tests within that suite whose name matches the pattern]
```

The last variant is the quickest way for me run one test repeatedly as I'm developing.  Because it runs everytime I save changes to disk, it often runs when my code is in some bad state and everything fails.  But no big deal, it just runs again when I fix things, so I ignore the window with the running tests until I think I have things in an OK state.

some more description of the arguments to scalatest itself (eg `-z` http://www.scalatest.org/user_guide/using_the_runner)

**Comment by jinxing64** [comment]
@squito 
Many thanks for your help.  You are so kind person : )

**Comment by kayousterhout** [comment]
I merged this but just realized the current JIRA number is wrong.  I won't do a revert / recommit because this is such a small PR, but in the future, each PR should have its own JIRA (so this should have had a separate JIRA about the scheduler test race condition).

**Comment by jinxing64** [comment]
@kayousterhout Thanks a lot. Sorry for this and I'll be careful in the future.

**Pull request closed without merging**